runtime/gc_structs/ClassLoaderClassesIterator.cpp

```diff
@@ -108,9 +108,22 @@ GC_ClassLoaderClassesIterator::nextClass()
 		if (ANONYMOUS_CLASSES == _mode) {
 			_nextClass = nextAnonymousClass();
 		} else {
-			if ( (result->classLoader == _classLoader) && (NULL != result->arrayClass) ) {
+#if defined(J9VM_OPT_VALHALLA_FLATTENABLE_VALUE_TYPES)
+			if ((result->classLoader == _classLoader) && (NULL != result->nullRestrictedArrayClass)) {
+				/* class has a null restricted array but no arrayClass */
+				_nextClass = result->nullRestrictedArrayClass;
+			} else
+#endif /* defined(J9VM_OPT_VALHALLA_FLATTENABLE_VALUE_TYPES) */
+			if ((result->classLoader == _classLoader) && (NULL != result->arrayClass)) {
 				/* this class is defined in the loader, so follow its array classes */
 				_nextClass = result->arrayClass;
+#if defined(J9VM_OPT_VALHALLA_FLATTENABLE_VALUE_TYPES)
+			} else if ((result->classLoader == _classLoader)
+				&& J9_IS_J9ARRAYCLASS_NULL_RESTRICTED(result)
```

Comment:
Do we need this condition? Current assumption is any slot can contain pointer to j9class should be scanned unconditionally except it is NULL.

Reply:
Yes, otherwise there could be an infinite loop. If there exists one array class of arity one and a nullRestrictedArrayClass object for a base ram class they will be pointing to each other as companion classes. 
```
   baseClass-->arrayClass
           |
           | --->nullRestrictedArrayClass
```
arrayClass->companionClass is nullRestrictedArraClass
nullRestrictedArrayClass->companionClass is arrayClass

Reply:
Please note that null restricted array can only be 1D array, but you can still have a 2D nullable array whose element is 1D null restricted array. So it is possible to have things like `A, A[], A[][], A[][][], A[!], A[][!], A[][][!]` (`[!]` means null is not allowed in that dimension and` [!]` can only be at the last dimension):
```

       |------->arrayClass-->arrayClass-->arrayClass
       |          A[]           A[][]       A[][]
       |                   
baseClass (A)     
       |                   
       |           A[!]                   A[][!]      A[][][!]
       |----> nullRestrictedArrayClass-->ArrayClass-->ArrayClass
```
Only `A[]` and` A[!]` points to each other as companions.

You need sth like:

```
J9Class* getNullRestrictedArray(J9Class* c) {
     J9Class* baseClass = NULL;
     if (NULL == c->leafComponentType) {
        /* c is a baseClass*/ 
        baseClass = c;
     } else {
         /* c is an arrayClass,  c->leafComponentType is the baseClass*/ 
        baseClass = c->leafComponentType;
     }
     return baseClass->nullRestrictedArraClass;
}

if ((result->classLoader == _classLoader) && (NULL != result->arrayClass)) {
	/* this class is defined in the loader, so follow its array classes */
	_nextClass = result->arrayClass;
else if ((result->classLoader == _classLoader) && (NULL != getNullRestrictedArray(result))) {
	_nextClass = getNullRestrictedArray(result);
} else if (TABLE_CLASSES == _mode) {
	_nextClass = nextTableClass();
} else {
	_nextClass = nextSystemClass();
}
```

Reply:
That looks good to me. I added your suggestion to the change.